Restricted choices feature

akanellop opened the pull request:
separate environmental value with type or restricted choices according to a constant

akanellop committed:
Restricted choices feature , check syntax for appearing [ ]
separate environmental value with type or restricted choices according to a constant

akanellop commented:
In fuction checkValidationOfValues, we check if there is syntax for restricted choices(if there are [ ] ) and pass an extra argument in element (= element[environmental variable,type/restrictedchoices,code])
so outputHTML function can print the right html string according to what follows the environmental variable(type or restricted value)
Added tests

parisk commented on src/main.test.js:
```diff
@@ -44,3 +57,8 @@ test("Multiple invalid variables and types", () => {
   expect(envSpecToHTML(testEnv3)).toEqual("Error:Wrong Syntax");
 });
+
+test("Wronf Syntax for restricted choices", () => {
```
Wronf → Wrong 😅

parisk commented on src/main.js:
```diff
@@ -79,4 +111,5 @@ const outputHTML = envValues => {
 const envSpecToHTML = envSpec => {
   return outputHTML(checkValidationOfValues(envSpec));
 };
+console.log(envSpecToHTML("VASAA : email"));
```
Remove this please. We do not want `console.log` statements in libraries 😄.

parisk commented on src/main.js:
```diff
@@ -22,11 +22,23 @@ const checkValidationOfValues = envSpecString => {
       validTypes.includes(element[1])
     ) {
+      //in case value has valid type
+      element[2] = 0; //0 constant that declares what should be printed for an element with type
```
Changing the value of the argument provided in a `filter` function (in our case this is `element`), is not a good practice. We should:

1. Find another way to store this information
2. We should prefer something more human-friendly than numbers (0, 1) to represent that kind of information

parisk commented on src/main.js:
```diff
@@ -22,11 +22,23 @@ const checkValidationOfValues = envSpecString => {
+    // in case value has restricted choices (indicated by "[]" ,we should split them
+    else if (
+      element[0].match(alphanumericThatDoesNotStartWithDigit) &&
+      element[1][0] === "[" &&
```
I see what you are doing here, but it's quite crude and involves multiple checks. We should prefer using a regular expression and extract the options inside:

- About Regular Expressions: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Regular_Expressions
- Extracting information with parentheses: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Regular_Expressions#Using_parenthesized_substring_matches

parisk commented on src/main.js:
```diff
@@ -59,15 +71,35 @@ const parseVarFromType = envSpecAsArray => {
+        case 1:
+          //if element is value with restricted choices
+          toPrint =
+            `<label for="env_spec_admin_${element[0].toLowerCase()}">${
```
Since the code for the output label is the same whether we have types, restricted options or nothing, we should extract this to a function that we reuse in both places.

parisk reviewed (changes requested):
I am really happy to see this working! I would like us to refine the code before considering merging 😄.

akanellop committed:
changed filter function to map  function in checkValidationOfValues
used string constants instead of int constansts for separating types and restricted choices
new generic expression for checking syntax for restricted choices

parisk commented on src/main.js:
```diff
@@ -16,17 +16,32 @@ const checkValidationOfValues = envSpecString => {
   ];
   const alphanumericThatDoesNotStartWithDigit = /^[A-Z_][0-9A-Z_]*$/;
+  const genericForCheckingRestrChoicesSyntax = /^\[(.*)\]$/;
   let envValArray = parseVarFromType(envSpecString.trim().split("\n")); //split lines based on \n character and parse them
```
Can you rename this to something more semantic? Now that I have to read more code, I struggle every time understanding what's inside this variable.

What about `envSpecLines` or something like that?

parisk commented on src/main.js:
```diff
@@ -16,17 +16,32 @@ const checkValidationOfValues = envSpecString => {
-  envValArray = envValArray.filter(element => {
-    //check for valid variables AND types
+  //element[2] will be what this environmental variable contains e.g. as default or as options
+  envValArray = envValArray.map(element => {
```
OK, we will have to make this human-friendlier by making a couple of changes:

1. Use a **new** _constant_ to store the result of `map` (e.g. `envSpecEntries`)
2. Use a structured object, instead of an array:
    ```javascript
    // Prefer this
    const entry = {
        name: 'ADMIN_EMAIL',
        type: 'test',  // If restricted choices are set, this should be `null`
        choices: null  // If restricted choices are set, this should be an array (e.g. `['a', 'b', 'c']`)
    }

    // Instead of this
    const entryInArrayFormat = ['ADMIN_EMAIL', 'email', null];

    console.log(entry.name);  // It's clear what we are printing here
    console.log(entry[0]);  // It's not clear what we are printing here
    ```

parisk commented on src/main.js:
```diff
@@ -59,15 +74,35 @@ const parseVarFromType = envSpecAsArray => {
-        }" />\n`
-    );
+    envValues = envValues.map(element => {
+      switch (element[2]) {
```
By implementing the structure object suggested above, this can be simplified to the following:

```javascript
if (entry.type) {
    // Do things for typed entries
} else {
    // Do things for restricted choices
}
```

parisk commented on src/main.js:
```diff
@@ -59,15 +74,35 @@ const parseVarFromType = envSpecAsArray => {
+      switch (element[2]) {
+        case "hasType": //if element is value with type
+          toPrint =
+            `<label for="env_spec_${element[0].toLowerCase()}">${
```
Since you are printing the same label for both typed entries and restricted choices, we should do one of the following:

1. Extracting this logic to a separate function (e.g. `renderLabelForEntry(entry)`)
2. Running this kind of logic outside the `if/switch` clause

akanellop replied:
I have a question, in your examples there is a different output for syntax with types and for syntax with choices.
In issue#3 https://github.com/sourcelair/env-spec-js/issues/3
input: ADMIN_EMAIL: email
output:
<label for="env_spec_admin_email">ADMIN_EMAIL</label>
<input id="env_spec_admin_email" name="admin_email" type="email" />

and in issue#4 https://github.com/sourcelair/env-spec-js/issues/4
input: DEBUG: [0,1]
output:
<label for="env_spec_admin_debug">DEBUG</label>
<select id="env_spec_admin_debug" name="admin_debug">
  <option value="0">0</option>
  <option value="1">1</option>
</select>

In the second example there is an extra admin word  in lines :<label for="env_spec_admin_debug">DEBUG</label> and <select id="env_spec_admin_debug" name="admin_debug">
did this happen by mistake?

parisk reviewed (changes requested):
I have some comments basically on code quality and variable namings and we are good to go!

akanellop committed:
-changed multidiamensional arrays so they contain objects with attributes
-new function that creates render label
-changed variables' names to be more explanatory

akanellop committed:
deleted not needed console log

parisk commented on src/main.js:
```diff
@@ -16,60 +17,108 @@ const checkValidationOfValues = envSpecString => {
+      } else if (element.choices) {
+        //if element is value with restricted choices
+        toPrint += `<select id="env_spec_${element.name.toLowerCase()}" name="${element.name.toLowerCase()}">\n`;
+        for (let i = 0; i < element.choices.length; i++) {
```
Change the `for` loop please to one without counters 😄:

```javascript
for (const choice of element.choices) {
    // Do things with choice, instead of `element.choices[i]`. Much cleaner!
}
```

akanellop committed:
changed iteration of choices in function outputHTML for restricted choices
